requester pays and other storage details in Asset and Item (#991)
radiantearth/stac-spec: extensions/storage/README.md

# Object Storage Extension Specification

- **Title: Object Storage**
- **Identifier: storage**

> - **Field Name Prefix: -**

Reviewer: ```suggestion
- **Field Name Prefix: storage**
```

- **Scope: Item, Assets**
- **Extension [Maturity Classification](../README.md#extension-maturity): Proposal**
- **Owner**: @davidraleigh

> This document explains the fields of the Object Storage Extension to a STAC Item and an Asset. This does not cover NFS solutions provided by PaaS cloud companies. Allows user to specify details related to cloud storage access and costs for assets and their STAC item.

Reviewer: ```suggestion
This document describes the Object Storage Extension.  This extension adds fields to STAC Item and Asset objects. This allows for details related to cloud storage access and costs for assets to be associated with a STAC Item.

This extension does not cover NFS solutions provided by PaaS cloud companies. 
```

- [Example (Landsat 8)](examples/example-landsat8.json)
- [JSON Schema](json-schema/schema.json)

## Asset fields

| Field Name  | Type   | Description |
| ----------- | ------ | ----------- |

> | storage:cloud_platform        | string    | The [cloud provider](#providers) where data is stored |

Reviewer 2: I agree the acceptable values here should be a well-defined enum. E.g. `aws` vs `amazon_web_services`

> | storage:cloud_platform        | string    | The [cloud provider](#providers) where data is stored |

Reviewer: ```suggestion
| storage:platform        | string    | The [cloud provider](#providers) where data is stored |
```

maybe just platform?

Reviewer 3: Yes, shorter is better and cloud seems superfluous

Author: I was avoiding platform because C++ and enums and platform clashing with stac item platform. But I can make it `platform` for this pull request and make it `storage_platform` in protobuf.

Author: https://github.com/radiantearth/stac-spec/pull/991/commits/09510ce3e91b73010a93991defdd51856c47c3b1

| storage:manager               | string    | The entity in charge of managing the data. |
| storage:regions               | [string]  | The region(s) where the data is stored. Relevant to speed of access and inter region egress costs (as defined by PaaS provider) |

> | storage:bucket                | string    | The bucket for the asset, used along with object path |

Reviewer 4: Azure doesn't have a single concept of a 'bucket', but has a 'storage account' and 'container'.

It seems like the `storage:bucket` and `storage:object_path` are redundant information for the asset `href`. Is the idea the `href` would be an http(s) url? I think if we were to drop the bucket and object_path properties, clients would be able to parse information from either the http endpoint or something like `s3://...`.

Reviewer 3: I'm not a big fan of the redundant information by having `bucket` and `path` since it is already in href which encapsulates not only bucket, path, but also protocol.

Author: We use STAC a lot internally, so object_path and bucket are useful to those internal users who have access permissions to use them, but for customers there is an href that isn't constructed from bucket + object_path.

https://github.com/davidraleigh/stac-spec/blob/issue-896/extensions/storage/examples/example-nsl.json#L66
https://github.com/davidraleigh/stac-spec/blob/issue-896/extensions/storage/examples/example-nsl.json#L69

> | storage:object_path           | string    | The object_path for the asset, used along with bucket |

Reviewer 5: Maybe just `path`?

Reviewer: > Maybe just path?

I think `storage:container` and `storage:path` are good names. With an explicit example, e.g., 

```
{ 
 "href": "https://foobar-data.s3.us-west-2.amazonaws.com/some/path/to/a/cog.tif",
  "storage:platform": "aws_s3",
  "storage:container": "foobar-data",
  "storage:path": "some/path/to/a/cog.tif",
  "storage:region":"us-west-2"
}
```

Author: > I think `storage:container` and `storage:path` are good names. With an explicit example, e.g.,
It would be strange to adopt Azure's `container` when all the other providers use the term `bucket`

`key` is a more common term used by cloud providers (in addition to `object` and `blob`). `path` isn't really anywhere in documentation. I just added it to `object` to make that more clear.

Reviewer 3: key is more descriptive as well, I think it's a better term than path

Reviewer 5: I'm fine either way as long as we avoid the `_` 😅. For people not so much into cloud storage path seems more intuitive, but if key is an accepted term across providers, then I guess it's fine.

Reviewer: bucket and key sound fine then, if all the other providers use it. I was trying to counter my AWS-bias and overcorrected.

> | storage:requester_pays        | bool      | Is the data requester pays or is it data manager/cloud provider pays |

Reviewer: requester_pays as a term is AWS specific, but I think it's also become the standard way of describing this

Reviewer 3: Yeah, since Google uses this as well, and it's also a very descriptive term I think it's good to use generally.  

If Azure implements a similar mechanism my guess is they will adopt the same term

Author: > If Azure implements a similar mechanism my guess is they will adopt the same term
they were the only one of the six cloud providers I investigated to buck the trend and use `container` instead of `bucket`...

Reviewer: oh, interesting -- I only use AWS, so I just assumed that other services didn't use bucket out of spite -- I'm fine with it if that's the common term.

| storage:tier                  | string    | The title for the tier type (as defined by PaaS provider) |

> | storage:tier_duration         | integer   | Minimum storage duration required before additional fees |

Reviewer: tier_duration seems like we're putting the cost structure schedule of the cloud provider into the metadata of the stac item. I'm not sure what the usefulness of this is.

Author: all cloud providers have a concept of a duration that a storage object is required to remain at that storage level before it can be retrieved without cost. It's not as precise as cost structure information, but it's similar to `requester_pays` in importance.

Reviewer: I get requester_pays, since that has to be set in the request or you can't retrieve the item. I don't know if that's ever used to make a decision about whether it should be retrieved or not (e.g., don't retrieve anything you have to pay for).  How would tier_duration be used? I just want to know what the use case is.

> | storage:date_stored           | string    | Date and time the corresponding asset placed into the current storage tier (relevant for tier_duration > 0) |

Reviewer: ```suggestion
| storage:date_stored           | string    | Date and time the corresponding asset placed into the current storage tier (relevant for tier_duration > 0). Format is RFC 3339. |
```

Reviewer 3: I thought we might have had a note on using the `created` and `updated` fields under Assets, but we don't:
https://github.com/radiantearth/stac-spec/blob/dev/best-practices.md#common-use-cases-of-additional-fields-for-assets

But you could use `created` under Assets which would imply when the asset was first created, and `updated` would be when it was last updated.  I think that covers this particular use, @m-mohr thoughts?

Reviewer 5: We have, but it's in this chapter: https://github.com/radiantearth/stac-spec/blob/dev/item-spec/common-metadata.md#date-and-time

Reviewer 5: And there's also the timestamps extension, which covers this probably better with "published"

> | storage:first_byte_latency    | string    | time unit (milliseconds, minutes or hours) for accessing first byte of data |

Reviewer: ```suggestion
| storage:first_byte_latency    | string    | magnitude for time for accessing first byte of data (e.g., `milliseconds`, `seconds`,  `minutes`, or `hours`) |
```

Reviewer 5: Could we make that a definitive list of options? (i.e. not "e.g." but "one of")

- milliseconds
- seconds
- minutes
- hours
- days
- weeks
- months
?

Reviewer: i think hours is probably the highest we need to go.

Reviewer 5: We have some companies that store on tape, which can take days to get the data from. Not sure this is covered by this extension though. And also not sure "first_byte_latency"  is the best term then

Reviewer 3: I'm not convinced of this field is needed at all.  Doesn't `tier` imply a certain latency for a given provider?  
As a provider how would you populate this field? Would you basically look up the latency for the tier you are storing your data and use that?  If so then I don't think it is needed.

If is is needed, than `latency` would seem to be sufficient.

Author: @matthewhanson tier isn't directly related. For Alibaba's "Cold Archive", there are different options for your speed to first byte depending on the fees you're willing to pay, 1 hour, 2 hours, etc. 

it's not exactly latency. it really is a delay of some period of time and then your latency to download is the normal access speed. Oracle and Microsoft describe it as "Time to first byte" and AWS uses "First Byte Latency".

Reviewer 3: I see, that makes sense. I don't have a strong opinion here, but agree with @m-mohr that this should be an enumerated list of options.

> 

Reviewer: ```suggestion
**storage:cloud_platform** value representing the name of the object store. Valid values for common cloud storage providers are `aws_s3`, `alibaba_oss`, `microsoft_azure_storage`, `google_cs`, `ibm_os`, and `oracle_os`
```

Reviewer 5: How can we make that in the future we don't need to change the spec if a new cloud service appears? The ones here sound a bit US centric.

Reviewer 5: Maybe better have a table below with common services and their identifier, but make that list non-exclusive?

Reviewer: Yes, agree -- I phrased that wrong in text. I meant that for these, providers, use these values -- but there can and will be more

Reviewer 3: Does it need to be `aws_s3` or is `s3` sufficient, where we can rely on established protocol names already used in URLs? 
e.g., s3://, gs://   

`s3` for AWS and `gs` for Google seem to be widely used, there just doesn't seem to be an accepted upon standard for Azure.

Reviewer 5: I think I'd also like to use the protocol names, although several providers implement s3. Do they all use the same protocol then?

Reviewer: I think implementing s3 just means you conform to the AWS S3 REST API that clients (such as the cli) use anyway. 

I'd prefer separating the scheme from the actual service being used, and to follow the general pattern of `{provider}_{service_name}` -- otherwise, we'd have to assume `aws` meant `s3` (what if they later introduce `s4`?), whereas the other more generic names are just like `cloud_storage` have to have the provider. So, consistency.

Also, the Asset#href might be an ARN (`arn:aws:s3:::bucket_name/key_name`) and then you're back to parsing the URI to find the service.

## Item fields

| Field Name  | Type   | Description |
| ----------- | ------ | ----------- |
| storage:min_tier_duration   | string  | Date and timestamp type describing the shortest time tier restrictions on access of assets. If `storage:archived_small` is false then thumbnails and other small overviews (implementer's discretion) are not included in `min_tier_duration` calculation |
| storage:max_tier_duration   | string  | Date and timestamp type describing the longest time tier restrictions on access of assets. |

> | storage:archived_small         | bool      | Should the thumbnails or other small data be assumed to be included in the `min_tier_duration` and `max_tier_duration` |

Reviewer 2: "small data" seems quite vague. If the `storage` extension is applied to each asset, this shouldn't be necessary, because you can just see if there are `storage` keys within the thumbnail asset?

Reviewer 3: agree here, how do I know what is to be considered small?

Author: we can remove `archived_small` and just put the responsibility on the maintainer of the STAC service to not include thumbnails that are in Standard storage in the calculation of the min_tier_duration

Author: or to include them if they want all items archived

Author: https://github.com/radiantearth/stac-spec/pull/991/commits/09510ce3e91b73010a93991defdd51856c47c3b1

> | storage:archived              | bool      | descriptor for whether the data is "properly" archived according to implementers discretion |

Reviewer: I don't understand what this is supposed to be from the description

Author: @philvarner I'll put more details in the readme. This is about being able to query all your STAC items to see which items need to be transitioned to an Archived state.
